packages/core/src/client.ts

```diff
@@ -245,7 +235,63 @@ export class Client extends AsyncEventEmitter<ManagerShardEventsMap> {
 			},
 		});
 
-		return promise;
+		try {
+			for await (const payload of on(this, GatewayDispatchEvents.GuildMembersChunk, { signal: controller.signal })) {
+				const [{ data }] = payload as MappedEvents[GatewayDispatchEvents.GuildMembersChunk];
+
+				if (data.nonce !== nonce) continue;
+
+				clearTimeout(timer);
+
+				yield {
+					members: data.members,
+					nonce,
+					notFound: data.not_found ?? null,
+					presences: data.presences ?? null,
+					chunkIndex: data.chunk_index,
+					chunkCount: data.chunk_count,
+				};
+
+				if (data.chunk_index >= data.chunk_count - 1) {
+					break;
+				} else {
+					timer = createTimer();
+				}
+			}
+		} catch (error) {
+			if (error instanceof Error && error.name === 'AbortError') {
+				throw new Error('Request timed out');
+			}
```

Comment:
We shouldn't do this, technically `controller.abort(new Error('Request timed out'));` should achieve the same effect, although I can see stack traces dying there... maybe create the error before the timer? 🤔 

We should handle the case for this block:
```js
for await (const _ of client.requestGuildMembersIterator({ guild_id: '...' })) {
	AbortSignal.abort().throwIfAborted();
}
```

If my memory on `for await...of` is still correct, throwing an error inside a `for...of` (or a `for await...of`) causes `yield` to throw an error, which in this case it'll call this `catch`, and this piece will overwrite the user's error, which is undesirable and not intended.

Reply:
I agree, we should just remove this entire catch block and let errors bubble up

Reply:
I had already sent this in Discord, but I'll send it here as well just for transparency

> We shouldn't do this, technically `controller.abort(new Error('Request timed out'));` should achieve the same effect

Doing this makes it throw an AbortError with its `cause` set to `new Error('Request timed out')`, hence why I was catching & rethrowing, e.g.:
![image](https://github.com/discordjs/discord.js/assets/42935195/92ec0640-9ce8-4a8c-9d8f-8a303f004ae1)

> If my memory on `for await...of` is still correct, throwing an error inside a `for...of` (or a `for await...of`) causes `yield` to throw an error, which in this case it'll call this `catch`, and this piece will overwrite the user's error, which is undesirable and not intended.

This does not appear to be the case:

![image](https://github.com/discordjs/discord.js/assets/42935195/ef5636eb-1c29-416e-85cd-f71e18d27540)
![image](https://github.com/discordjs/discord.js/assets/42935195/cfa0c613-5377-4c55-a1fc-c15a858ee68a)

Reply:
After some internal discussion, we're keeping this as it is